File: 117. Populating Next Right Pointers in Each Node II/connect.js
// Definition for a _Node.
class _Node {
  constructor(val = null, left = null, right = null, next = null) {
    this.val = val;
    this.left = left;
    this.right = right;
    this.next = next;
  }
};

/**
 * @param {_Node} root
 * @return {_Node}
 */
const connect = (root) => {
  if (!root) {
    return root;
  }

  const queue = [root];

  while (queue.length) {
    let queueSize = queue.length;

    for (let i = 0; i < queueSize; i++) {
      const node = queue.shift();

      if (i < queueSize - 1) {
        node.next = queue[0];
      } else {
        node.next = null;
      }

      if (node.left) {
        queue.push(node.left);
      }

      if (node.right) {
        queue.push(node.right);
      }
    }
  }

  return root;
};


const root = new _Node(1);
root.left = new _Node(2);
root.right = new _Node(3);

root.left.left = new _Node(4);
root.left.right = new _Node(5);

root.right.right = new _Node(7);

console.log('connect: ', connect(root));

File: 117. Populating Next Right Pointers in Each Node II/connect.py
# Definition for a Node.
from collections import deque

class Node:
  def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
    self.val = val
    self.left = left
    self.right = right
    self.next = next

class Solution:
  def connect(self, root: 'Node') -> 'Node':
    if not root:
      return root

    q = deque([root])
    
    while q:
      qSize = len(q)
      for i in range(qSize):
        node = q.popleft()
        
        if i < qSize - 1:
          node.next = q[0]
        else:
          node.next = None
          
        if node.left:
          q.append(node.left)
          
        if node.right:
          q.append(node.right)
    
    return root
        

root = Node(1)
root.left = Node(2)
root.right = Node(3)

root.left.left = Node(4)
root.left.right = Node(5)

root.right.right = Node(7)

solution = Solution()
print('connect: ', solution.connect(root))
